cps/kni.c

```diff
@@ -911,32 +911,52 @@ rd_send_batch(struct cps_config *cps_conf, struct mnl_nlmsg_batch *batch,
 {
 	/* Address of routing daemon. */
 	struct sockaddr_nl rd_sa;
+	bool send_done_after = false;
 	int ret = 0;
 
 	memset(&rd_sa, 0, sizeof(rd_sa));
 	rd_sa.nl_family = AF_NETLINK;
 	rd_sa.nl_pid = pid;
 
-	if (done) {
-		struct nlmsghdr *done =
-			mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
-		done->nlmsg_type = NLMSG_DONE;
-		done->nlmsg_flags = NLM_F_MULTI;
-		done->nlmsg_seq = seq;
-		done->nlmsg_pid = cps_conf->nl_pid;
-	}
+	while (1) {
+		if (done) {
+			struct nlmsghdr *done = mnl_nlmsg_put_header(
+				mnl_nlmsg_batch_current(batch));
+			done->nlmsg_type = NLMSG_DONE;
+			done->nlmsg_flags = NLM_F_MULTI;
+			done->nlmsg_seq = seq;
+			done->nlmsg_pid = cps_conf->nl_pid;
+			if (!mnl_nlmsg_batch_next(batch)) {
```

Comment:
Having the following code as the body of this `if` statement just sould simplify the code:
```
/* Sent the *full* batch out without the DONE message. */
ret = rd_send_batch(cps_conf, batch, daemon, seq, pid, false);
if (ret < 0)
	return ret;
/* Go on to sent the the DONE message. */
```

This change will avoid the `while` loop, variable `send_done_after`, and change the code outside of the `if (done) {`.

Reply:
I think I'm still a little confused. In my view, we can send the full batch without the DONE message as you described, but after the return of this call to `rd_send_batch()` with `done = false`, the batch will be empty. Since the first attempt to add the DONE message failed, we need to execute the code to add it to the batch again.

Reply:
Hence the `while` loop in my latest rebase.

Reply:
Another way of stating it: after we send the full batch (without the DONE) message, `mnl_nlmsg_batch_size(batch)` will be 0 in the call to `sendto()` so no data will be sent.

Reply:
Let's discuss this in our meetup today.

Reply:
I just rediscovered that the batching API takes care of overflow for you, so we don't have to handle it. That was the missing piece.